llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -1589,80 +1610,121 @@ bool SimplifyCFGOpt::hoistCommonCodeFromSuccessors(BasicBlock *BB,
   }
 
   bool Changed = false;
+  auto *SuccIterPairBegin = SuccIterPairs.begin();
+  SuccIterPairBegin++;
+  auto OtherSuccIterPairRange =
+      iterator_range(SuccIterPairBegin, SuccIterPairs.end());
+  auto OtherSuccIterRange = make_first_range(OtherSuccIterPairRange);
+  using InstrFlagPair = std::pair<Instruction *, unsigned>;
+  SmallVector<DenseMap<llvm::hash_code, InstrFlagPair>, 2> OtherSuccessorsHash;
+
+  for (auto BBItrPair : OtherSuccIterRange) {
+    // Fill the hashmap for every other successor
+    DenseMap<llvm::hash_code, InstrFlagPair> hashMap;
```

Comment:
Should this work directly with `Instruction`?
```suggestion
    DenseMap<Instruction *, InstrFlagPair> hashMap;
```

Reply:
I assume you are going to use this hash value to look up the same instructions for other BB's? But there may be different instructions with the same hash in a BB.

Reply:
That's correct, in some cases (not often though) it might happen so I when I look up an instruction from hashmaps, I also check it to be identical to the instruction I'm hoisting with.

Reply:
> That's correct, in some cases (not often though) it might happen so I when I look up an instruction from hashmaps, I also check it to be identical to the instruction I'm hoisting with.

Hmm, sorry, I am referring to that when creating this map, we may lose the same hash value with different instructions. Maybe you could try something like this `Map<SomeHash, Vec<Instruction *>>`? You can also quickly skip some instructions directly by the size of the `Vec`. Perhaps you can find a similar approach at https://llvm.org/docs/ProgrammersManual.html.

Reply:
I see. Right I can keep a vector of instructions for each hash. I just didn't understand the part about skipping instructions by the size of the vector. Could you please explain a bit more?

Reply:
After switching to a vector, if we find here that the number of instructions is less than the number of successors, we can just skip it. See https://github.com/llvm/llvm-project/pull/78615/files#diff-f1d6bccbcc778d09cdbe3f6f9d215299c9eaf9c9af6f66dbe6c90c4ad8e53904R1717-R1724.

Reply:
Thanks, I got your point.
With the current implantation we might miss some coroner cases as you mentioned however, it will happen rarely.
I'll be happy to make the changes you suggested to cover those cases but I guess it might result in some compile time overhead. 
If I have more than one instruction from a BB in a vector, then I need to go over each instruction in that vector and decide which ones should be hoisted. If all those instructions are identical, this decision and checks it requires might become a bit expensive however, if you think it is necessary, I'll be glad to implement that.

Reply:
I think this might improve compile times, if we ignore hash collisions.
If the hash is the same, we overwrite or ignore the previous value in a vector. Then you could just check that the length of the vector matches the number of switch cases. This may reduce a lot of hash computations.
I'm not sure whether it works, but I think it's worth a try.

Reply:
Sorry for the delay. You're right, ignoring collisions is helpful. I made the changes as you suggested.